doc/dwb.1
'\" t
.\"     Title: dwb
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.77.1 <http://docbook.sf.net/>
.\"      Date: 01/27/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "DWB" "1" "01/27/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
dwb \- dynamic web browser
.SH "SYNOPSIS"
.sp
\fBdwb\fR [\fIOPTIONS\fR] [ \fIURLS\fR ]
.SH "DESCRIPTION"
.sp
dwb(1) is a small webbrowser based on WebKit and GTK which aims to be mostly keyboard\-driven\&.
.SH "OPTIONS"
.PP
\fB\-e, \-\-embed\fR=\fIwid\fR
.RS 4
Embed dwb into <wid>\&.
.RE
.PP
\fB\-f, \-\-force\fR
.RS 4
Force restoring a previously saved session, even if it seems that another process has opened a session with the same name\&.
.RE
.PP
\fB\-l, \-\-list\-sessions\fR
.RS 4
List previously saved sessions\&. A
\fI*\fR
indicates that another instance has currently opened the session\&.
.RE
.PP
\fB\-n, \-\-new\-instance\fR
.RS 4
New instance, overrides setting
\fIsingle\-instance\fR\&.
.RE
.PP
\fB\-p, \-\-profile\fR=\fIprofilename\fR
.RS 4
Load configuration for profile
\fIprofilename\fR\&.
.RE
.PP
\fB\-r, \-\-restore\fR=\fIsessionname\fR
.RS 4
Restore session with name
\fIsessionname\fR
or the default session, if
\fIsessionname\fR
is omitted\&.
.RE
.PP
\fB\-R, \-\-override\-restore\fR
.RS 4
Open a new session, even if
\fIsave\-session\fR
is enabled\&.
.RE
.PP
\fB\-S, \-\-enable\-scripts\fR
.RS 4
Enables the javascript interface, if there are scripts in $XDG_CONFIG_HOME/dwb/userscripts that use the javascript api this option is not needed\&.
.RE
.PP
\fB\-x, \-\-execute\fR=\fIcommands\fR
.RS 4
Execute a list of dwb commands separated by
\fI;;\fR, see
\fICUSTOM COMMANDS\fR
and
\fICOMMAND OVERVIEW\fR
for details\&. If
\fIsingle instance\fR
is enabled all commands will be executed in the primary instance\&.
.RE
.PP
\fB\-v, \-\-version\fR
.RS 4
Show version information and exit\&.
.RE
.SH "MODES"
.sp
dwb has different modes:
.PP
\fBNormal mode\fR
.RS 4
The default mode\&. Pressing
\fIEscape\fR
always enter normal mode, all keyboard inputs will be interpreted as shortcuts, keypresses are not sent to the webview\&.
.RE
.PP
\fBInsert mode\fR
.RS 4
Used for editing text elements in a webpage\&.
.RE
.PP
\fBHint mode\fR
.RS 4
Follow links via hints\&.
.RE
.PP
\fBCommand mode\fR
.RS 4
Execute dwb commands from the builtin commandline\&.
.RE
.SH "SHORTCUTS"
.SS "Normal mode shortcuts"
.PP
\fB[n]j\fR
.RS 4
Scroll [n times] down (command:
\fIscroll_down\fR, aliases:
\fIdown\fR)\&.
.RE
.PP
\fB[n]k\fR
.RS 4
Scroll [n times] up (command:
\fIscroll_up\fR, aliases:
\fIup\fR)\&.
.RE
.PP
\fB[n]h\fR
.RS 4
Scroll [n times] left (command:
\fIscroll_left\fR, aliases:
\fIleft\fR)\&.
.RE
.PP
\fB[n]l\fR
.RS 4
Scroll [n times] right (command:
\fIscroll_right\fR, aliases:
\fIright\fR)\&.
.RE
.PP
\fB[n]gg\fR
.RS 4
Scroll to the top or to n% of the page (command:
\fIscroll_top\fR, aliases:
\fItop\fR)\&.
.RE
.PP
\fB[n]G\fR
.RS 4
Scroll to the bottom or to n% of the page (command:
\fIscroll_bottom\fR, aliases:
\fIbottom\fR)\&.
.RE
.PP
\fB[n]C\-f\fR
.RS 4
Scroll [n] pages down (command:
\fIscroll_page_down\fR, aliases:
\fIpagedown\fR)\&.
.RE
.PP
\fB[n]C\-b\fR
.RS 4
Scroll [n] pages up (command:
\fIscroll_page_up\fR, aliases:
\fIpageup\fR)\&.
.RE
.PP
\fB[n]C\-d\fR
.RS 4
Scroll [n] half pages down (command:
\fIscroll_halfpage_down\fR, aliases:
\fIhalfdown\fR)\&.
.RE
.PP
\fB[n]C\-u\fR
.RS 4
Scroll [n] half pages up (command:
\fIscroll_halfpage_up\fR, aliases:
\fIhalfup\fR)\&.
.RE
.PP
\fB:\fR
.RS 4
Enter Command mode\&.
.RE
.PP
\fBo\fR
.RS 4
Open new url in the focused tab, the argument $URI will be replaced by the current uri (command:
\fIopen\fR, aliases:
\fIo\fR)\&.
.RE
.PP
\fBgo\fR
.RS 4
Open url in the focused tab, set current url in the navigation bar (command:
\fIopen_url\fR)\&.
.RE
.PP
\fBO\fR
.RS 4
Open url in a new tab, the argument $URI will be replaced by the current uri (command:
\fItabopen\fR, aliases:
\fItopen\fR,
\fIt\fR)\&.
.RE
.PP
\fBgO\fR
.RS 4
Open url in a new tab, set current url in the navigation bar (command
\fItabopen_url\fR)\&.
.RE
.PP
\fBxo\fR
.RS 4
Open url in a new tab in background, the argument $URI will be repaced with the current uri (command:
\fIbackopen\fR, aliases:
\fIbopen\fR
)\&.
.RE
.PP
\fBxO\fR
.RS 4
Open url in a new tab in background, set current url in the navigation bar (command:
\fIbackopen_url\fR)\&.
.RE
.PP
\fBwo\fR
.RS 4
Open url in a new instance (command:
\fIwinopen\fR, aliases:
\fIwopen\fR,
\fIw\fR)\&.
.RE
.PP
\fBi\fR
.RS 4
Enter insert mode (command:
\fIinsert_mode\fR, aliases:
\fIi\fR,
\fIinsert\fR)\&.
.RE
.PP
\fBC\-n\fR
.RS 4
Enter normal mode\&.
.RE
.PP
\fBga\fR
.RS 4
Add a new blank tab (command:
\fItab_new\fR)\&.
.RE
.PP
\fBd\fR
.RS 4
Close current tab (command:
\fIclose_tab\fR, aliases:
\fIclose\fR)\&.
.RE
.PP
\fB[n]gc\fR
.RS 4
Clear tab n or of current tab, clears the history of the tab and loads
\fIabout:blank\fR\&. (command
\fIclear_tab\fR, aliases:
\fIclear\fR)\&.
.RE
.PP
\fBco\fR
.RS 4
Close all tabs except for the current one (command
\fIonly\fR)\&.
.RE
.PP
\fBC\-q\fR
.RS 4
Quit (command
\fIquit\fR, aliases:
\fIq\fR)\&.
.RE
.PP
\fBu\fR
.RS 4
Undo closing last tab (command
\fIundo\fR, aliases:
\fIu\fR)\&.
.RE
.PP
\fBC\-h\fR
.RS 4
Open the default startpage (command
\fIstart_page\fR, aliases:
\fIhome\fR)\&.
.RE
.PP
\fBH\fR
.RS 4
Go back (command
\fIhistory_back\fR, aliases:
\fIback\fR,
\fIba\fR)\&.
.RE
.PP
\fBL\fR
.RS 4
Go forward (command
\fIhistory_forward\fR, aliases:
\fIforward\fR,
\fIfo\fR)\&.
.RE
.PP
\fBth\fR
.RS 4
Go back in a new tab (command
\fItab_hist_back\fR, aliases:
\fItabback\fR,
\fItba\fR)\&.
.RE
.PP
\fBtl\fR
.RS 4
Go forward in a new tab (command
\fItab_hist_forward\fR, aliases:
\fItabforward\fR,
\fItfo\fR)\&.
.RE
.PP
\fBwh\fR
.RS 4
Go back in a new window (command
\fIwin_hist_back\fR, aliases:
\fIwinback\fR,
\fIwba\fR)\&.
.RE
.PP
\fBwl\fR
.RS 4
Go forward in a new window (command
\fIwin_hist_forward\fR, aliases:
\fIwinforward\fR,
\fIwfo\fR)\&.
.RE
.PP
\fB/\fR
.RS 4
Find forward (command
\fIfind_forward\fR, alias:
\fIffind\fR)\&.
.RE
.PP
\fB?\fR
.RS 4
Find backward (command
\fIfind_backward\fR, alias
\fIbfind\fR)\&.
.RE
.PP
\fBc/\fR
.RS 4
Find forward, ignore case (command
\fIfind_forward_ic\fR, alias:
\fIiffind\fR)\&.
.RE
.PP
\fBc?\fR
.RS 4
Find backward, ignore case (command
\fIfind_backward_ic\fR, alias
\fIibfind\fR)\&.
.RE
.PP
\fBM\fR
.RS 4
Add a bookmark (command
\fIbookmark\fR, aliases:
\fIbmark\fR,
\fIbma\fR)\&.
.RE
.PP
\fBgb\fR
.RS 4
Show Bookmarks (command
\fIbookmarks\fR, aliases:
\fIbmarks\fR,
\fIbmas\fR)\&.
.RE
.PP
\fBgB\fR
.RS 4
Show Bookmarks, open bookmark in a new tab (command
\fItab_bookmarks\fR, aliases:
\fItabmarks\fR)\&.
.RE
.PP
\fBwB\fR
.RS 4
Show Bookmarks, open bookmark in a new window (command
\fIwin_bookmarks\fR, aliases:
\fIwinmarks\fR)\&.
.RE
.PP
\fBm\fR
.RS 4
Add a quickmark (command
\fIsave_quickmark\fR, aliases:
\fIquickmark\fR,
\fIqmark\fR)\&.
.RE
.PP
\fBb\fR
.RS 4
Open quickmark (command
\fIquickmark\fR, aliases:
\fIqmarks\fR)\&.
.RE
.PP
\fBB\fR
.RS 4
Open quickmark in a new tab (command
\fItab_quickmark\fR, aliases:
\fItabqmarks\fR)\&.
.RE
.PP
\fBwb[:graph:]\fR
.RS 4
Open quickmark in a new window (command
\fIwin_quickmark\fR, aliases:
\fIwinqmarks\fR)\&.
.RE
.PP
\fB[n]r\fR
.RS 4
Reload tab n or current tab if n is omitted (command
\fIreload\fR, aliases:
\fIre\fR)\&.
.RE
.PP
\fB[n]R\fR
.RS 4
Reload tab n or current tab if n is omitted without using any cached data (command
\fIreload_bypass_cache\fR, aliases:
\fIrefull\fR)\&.
.RE
.PP
\fB[n]C\-s\fR
.RS 4
Stop loading of tab n or of current tab is [n] is omitted (command
\fIstop_loading\fR, aliases:
\fIstop\fR,
\fIst\fR)\&.
.RE
.PP
\fB[n]+\fR
.RS 4
Zoom in [n times] (command
\fIzoom_in\fR, aliases:
\fIzi\fR)\&.
.RE
.PP
\fB[n]\-\fR
.RS 4
Zoom out [n times] (command
\fIzoom_out\fR, aliases:
\fIzo\fR)\&.
.RE
.PP
\fB[n]=\fR
.RS 4
Zoom to n percent or to 100% if n is omitted(command
\fIzoom\fR, aliases:
\fIz\fR)\&.
.RE
.PP
\fBsf\fR
.RS 4
Save all configuration files (command
\fIsave\fR)\&.
.RE
.PP
\fBZZ\fR
.RS 4
Save current session and exit (command
\fIsave_session\fR, aliases:
\fIwq\fR)\&.
.RE
.PP
\fBgZZ\fR
.RS 4
Save current session with name and exit (command
\fIsave_named_session\fR, aliases:
\fIwqn\fR)\&.
.RE
.PP
\fB[n]J\fR
.RS 4
Cycle focus [n tab] forwards\&. (command
\fIfocus_next\fR, aliases:
\fItabnext\fR)\&.
.RE
.PP
\fB[n]K\fR
.RS 4
Cycle focus [n tab] backwards\&. (command
\fIfocus_prev\fR, aliases:
\fItabprev\fR)\&.
.RE
.PP
\fB[n]T\fR
.RS 4
Focus nth tab, first tab if n is omitted or last tab if n is 0\&. (command
\fIfocus_tab\fR, aliases:
\fItab\fR)\&.
.RE
.PP
\fBC\-Tab\fR
.RS 4
Toggle between current and last focused tab (command
\fItoggle_tab\fR, aliases:
\fIttab\fR)
.RE
.PP
\fB[n]gm\fR
.RS 4
Move current tab to position [n] or to first position if n is omitted\&. (command
\fItab_move\fR, aliases:
\fItabm\fR)\&.
.RE
.PP
\fB[n]gl\fR
.RS 4
Move current tab [n] positions left\&. (command
\fItab_move_left\fR, aliases:
\fItabl\fR)\&.
.RE
.PP
\fB[n]gr\fR
.RS 4
Move current tab [n] positions right\&. (command
\fItab_move_right\fR, aliases:
\fItabr\fR)\&.
.RE
.PP
\fBgt\fR
.RS 4
Show all open tabs\&. (command
\fIbuffers\fR, aliases:
\fIbu\fR)\&.
.RE
.PP
\fB[n]C\-P\fR
.RS 4
Protect tab [n]\&. Closing this tab must be confirmed (command
\fIprotect\fR, aliases:
\fIprot\fR)\&.
.RE
.PP
\fB[n]xd\fR
.RS 4
Lock tab [n]\&. Locking a tab will lock this tab to the current domain, it\(cqs not possible to navigate to another domain until unlocked\&.
\fIlock_domain\fR, aliases:
\fIdlock\fR)\&.
.RE
.PP
\fB[n]xu\fR
.RS 4
Lock tab [n]\&. Locking a tab will lock this tab to the current uri, it\(cqs not possible to navigate to another uri until unlocked\&.
\fIlock_uri\fR, aliases:
\fIulock\fR)\&.
.RE
.PP
\fBf\fR
.RS 4
Show hints (command
\fIhints\fR, aliases:
\fIhints\fR,
\fIhi\fR)\&.
.RE
.PP
\fBF\fR
.RS 4
Show hints, open link in a new foreground tab\&. (command
\fIhints_tab\fR, aliases:
\fItabhints\fR,
\fIthi\fR)\&.
.RE
.PP
\fB;b\fR
.RS 4
Show hints, open link in a new background tab\&. (command
\fIhints_background\fR, aliases:
\fIbackhints\fR,
\fIbhi\fR)\&.
.RE
.PP
\fBwf\fR
.RS 4
Show hints, open link in a new tab\&. (command
\fIhints_win\fR, aliases:
\fIwinhints\fR,
\fIwhi\fR)\&.
.RE
.PP
\fB;i\fR
.RS 4
Follow image (command
\fIhints_images\fR, aliases:
\fIihints\fR,
\fIihi\fR)\&.
.RE
.PP
\fB;I\fR
.RS 4
Follow image in a new tab (command
\fIhints_images_tab\fR, aliases:
\fIitabhints\fR,
\fIithi\fR)\&.
.RE
.PP
\fB\&.i\fR
.RS 4
Follow image in a background tab (command
\fIhints_images_background\fR, aliases:
\fIibackhints\fR)\&.
.RE
.PP
\fB;e\fR
.RS 4
Focus editable elements via hints (command
\fIhints_editable\fR, aliases:
\fIehints\fR,
\fIehi\fR)\&.
.RE
.PP
\fB;o\fR
.RS 4
Set hint\(cqs url in commandline (command
\fIhints_url\fR, aliases:
\fIuhints\fR,
\fIuhi\fR)\&.
.RE
.PP
\fB;O\fR
.RS 4
Set hint\(cqs url in commandline, open in a new tab (command
\fIhints_url_tab\fR, aliases:
\fIutabhints\fR,
\fIuthi\fR)\&.
.RE
.PP
\fB\&.o\fR
.RS 4
Set hint\(cqs url in commandline, open in a background tab (command
\fIhints_url_background\fR, aliases:
\fIubackhints\fR)\&.
.RE
.PP
\fB;d\fR
.RS 4
Download via hints (command
\fIhints_download\fR, aliases:
\fIdhints\fR)\&.
.RE
.PP
\fB;y\fR
.RS 4
Save link location to clipboard (command
\fIhints_clipboard\fR, aliases:
\fIchints\fR,
\fIchi\fR)\&.
.RE
.PP
\fB;Y\fR
.RS 4
Save link location to primary selection (command
\fIhints_primary\fR, aliases:
\fIphints\fR,
\fIphi\fR)\&.
.RE
.PP
\fB;r\fR
.RS 4
Rapid hint mode, each matching hint opens a new tab in background\&. (command
\fIhints_rapid\fR, aliases:
\fIrhints\fR,
\fIrhi\fR)\&.
.RE
.PP
\fB;R\fR
.RS 4
Rapid hint mode, each matching hint opens a new window\&. (command
\fIhints_rapid_win\fR, aliases:
\fIwrhints\fR,
\fIwrhi\fR)\&.
.RE
.PP
\fBgf\fR
.RS 4
Toggle "view source" (command
\fIview_source\fR, aliases:
\fIsource\fR,
\fIso\fR)\&.
.RE
.PP
\fBCC\fR
.RS 4
Allow persistent cookie for the current website\&. The domain will be saved in
\fIcookies\&.allow\fR\&. Cookies that are allowed by the cookies\&.allow whitelist are stored in $XDG_CONFIG_HOME/dwb/$profilename/cookies\&. (command
\fIallow_cookie\fR, aliases:
\fIcookie\fR)\&.
.RE
.PP
\fBCS\fR
.RS 4
Allow session cookie for the current website\&. The domain will be saved in
\fIcookies_session\&.allow\fR\&. This is only useful if
\fIcookies\-store\-policy\fR
is set to
\fInever\fR, see cookies\-store\-policy for details\&. (command
\fIallow_session_cookie\fR, aliases:
\fIscookie\fR)\&.
.RE
.PP
\fBCT\fR
.RS 4
Allow session cookies for the current website temporarily\&. Only first party cookies are allowed\&. The domain is not saved to a whitelist and the cookies will not be saved persitently\&. (command
\fIallow_session_cookie_tmp\fR, aliases:
\fItcookie\fR)\&.
.RE
.PP
\fB[n]yy\fR
.RS 4
Yank the url of tab n or of current tab if n is omitted to clipboard (command
\fIyank\fR)\&.
.RE
.PP
\fByY\fR
.RS 4
Yank the url of tab n or of current tab if n is omitted to primary selection (command
\fIyank_primary\fR, aliases:
\fIpyank\fR)\&.
.RE
.PP
\fByt\fR
.RS 4
Yank the title of tab n or of current tab if n is omitted to clipboard (command
\fIyank_title\fR, aliases:
\fItyank\fR)\&.
.RE
.PP
\fByT\fR
.RS 4
Yank the title of tab n or of current tab if n is omitted to primary selection (command
\fIyank_title_primary\fR, aliases:
\fItpyank\fR)\&.
.RE
.PP
\fBpp\fR
.RS 4
Paste from clipboard (command
\fIpaste\fR)\&.
.RE
.PP
\fBpP\fR
.RS 4
Paste from primary selection (command
\fIpaste_primary\fR, aliases:
\fIppaste\fR)\&.
.RE
.PP
\fBPp\fR
.RS 4
Paste from clipboard and load in a new tab (command
\fItab_paste\fR, aliases:
\fItpaste\fR)\&.
.RE
.PP
\fBPP\fR
.RS 4
Paste from primary selection and load in a new tab (command
\fItab_paste_primary\fR, aliases:
\fItppaste\fR)\&.
.RE
.PP
\fBwp\fR
.RS 4
Paste from clipboard and load in a new window (command
\fIpaste_nw\fR, aliases:
\fIwinpaste\fR)\&.
.RE
.PP
\fBwP\fR
.RS 4
Paste from primary selection and load in a new window (command
\fIpaste_primary_nw\fR, aliases:
\fIwinppaste\fR)\&.
.RE
.PP
\fB[n]ad\fR
.RS 4
Cancel the download with number n or the first download in the lists of running downloads if n is omitted\&. (command
\fIcancel_download\fR)\&.
.RE
.PP
\fBgs\fR
.RS 4
Add a searchengine\&. Textfields can be chosen with tab and a keyword must be specified\&. The first defined searchengine will be the default searchengine\&. The keyword can be used in all open commands, e\&.g\&.
\fI:open <keyword> <searchterm>\fR
(command
\fIsave_search_field\fR, aliases:
\fIsearch\fR)\&.
.RE
.PP
\fBgd\fR
.RS 4
Download the current site\&. (command
\fIdownload\fR)\&.
.RE
.PP
\fBSb\fR
.RS 4
Show bookmarks (command
\fIshow_bookmarks\fR, aliases:
\fIsbookmarks\fR)\&.
.RE
.PP
\fBSq\fR
.RS 4
Show quickmarks (command
\fIshow_quickmarks\fR, aliases:
\fIsquickmarks\fR)\&.
.RE
.PP
\fBSh\fR
.RS 4
Show history (command
\fIshow_history\fR, aliases:
\fIshistory\fR)\&.
.RE
.PP
\fBSd\fR
.RS 4
Show download (command
\fIshow_downloads\fR, aliases:
\fIsdownloads\fR)\&.
.RE
.PP
\fBSk\fR
.RS 4
Show keys (command
\fIshow_keys\fR, aliases:
\fIskeys\fR)\&.
.RE
.PP
\fBSs\fR
.RS 4
Show settings (command
\fIshow_settings\fR, aliases:
\fIssettings\fR)\&.
.RE
.PP
\fBss\fR
.RS 4
Set setting, the interactive version of the command set, for changing settings in scripts use set instead (command
\fIset_setting\fR)\&.
.RE
.PP
\fBsl\fR
.RS 4
Set local setting, changes a setting but doesn\(cqt save the setting to configuration file\&. The interactive version of the command local_set, for changing settings locally in scripts use local_set instead (command
\fIset_local_setting\fR)\&.
.RE
.PP
\fBsk\fR
.RS 4
Set keyboard shortcut (command
\fIset_key\fR, aliases:
\fIkeys\fR)\&.
.RE
.PP
\fBC\-p\fR
.RS 4
Toggle proxy (command
\fIproxy\fR)\&.
.RE
.PP
\fBtsh\fR
.RS 4
Toggle scripts for current host permanently (command
\fItoggle_scripts_host\fR, aliases:
\fIhscript\fR)\&.
.RE
.PP
\fBtsu\fR
.RS 4

CHANGELOG
Changes since 2012-10-28: 

New features:
 * Paste primary selection into webview
 * New option 'hints-key-lock'
 * Tab numbers in the status line
 * Extended javascript api

Bug fixes: 
 * Fixing :search command
 * Increasing performance of adblocker after a decrease with webkit-1.10
 * Fixing 2 segfaults in js-api
 * Ignore the default action for Mouse Buttons 8 and 9
 * Fixing a memory leak when closing a tab
 * Fixing segfault if userscript is a symlink
 * Ellipsize dwb:-links correctly
 * Update status bar if a tab is closed
 * Fixing wrong excaping when backslash is mapped to a shortcut

Removed: 
 * Command line option -s removed


Changes since 2012-08-22:

New features:
 * Indicator for quick/bookmarks
 * Scratchpad for userscripts
 * Update statusbar from userscripts
 * New option 'enable-webgl'
 * New option 'enable-accelerated-compositing'
 * New option 'show-single-tab'
 * Implementing 'mouseMove' signal for userscripts
 * Implementing 'Object.setPrivate', 'Object.getPrivate' for userscripts
 * Allow usage of directories for 'adblock-filterlist'
 * Extension manager for javascript extensions
 * New option 'sync-files', old 'sync-files' is now 'file-sync-interval'
 * New methods 'Array.fastIndexOf', 'Array.fastLastIndexOf'
 * More information in version info

Bug fixes:
 * Fixing element hider issues with webkit-1.10
 * Unescape uris with javascript-scheme
 * Always unprotect callback function in request_callback
 * Protect callback function in global_send_request
 * Adding missing download command to keys
 * Fixing wrong section number in dwb-js-manpage
 * Create js-quark before creating any objects; don't allocate position string in dwb_update_status_text
 * Use sizeof for stack-allocated strings; fixing one possible buffer overflow
 * Set default value for sync_files
 * Don't ignore case in directory listings
 * Create a new webview in view_create_web_view_cb if tabs are disabled
 * Fixing possible segfault in dwb_normal_mode
 * Let wv, frame and download be derived from gobject; fixing possible segfault in tabs_create; apply scripts before checking signals
 * Fixing several typos 
 * Ignore empty key-names in dwb:keys
 * Descriptions for cookie-functions
 * Respect evironment LDFLAGS
 * Check if key or value is NULL in dwb_set_key, fixes a possible segfault

Changes since 2012-05-11 

New features: 
 * A javascript api
 * Extensions: 'formfiller', 'requestpolicy', 'perdomainsettings', 'userscripts'
 * Token matching for tab completion
 * New command 'sanitize'
 * New option 'addressbar-dns-lookup'
 * New option 'close-last-tab-policy'
 * Create hints for image maps
 * Added DWB_PROXY environment variable to external download command
 * Use up/down key for tab completion
 * Adding download-speed to the download label
 * Focus hints with up/down-keys
 * New command 'toggle_tab'
 * New setting 'do-not-track'
 * Tilde-expansion for paths
 * New option 'bars-padding'
 * New option 'searchengine-submit-pattern'
 * New options 'progressbar-empty-color', 'progressbar-full-color',
   'progressbar-style'
 
Removed: 
 * $XDG_CONFIG_HOME/dwb/scripts/ is no longer supported, scripts from that
   directory are only injected with the 'userscripts' extension.
 * Removed command 'reload_scripts'.

Bug fixes:
 * Fixing wrong tab marked as protected when session is restored
 * Try to avoid overlapping hints
 * Fixing a bug when a new searchengine is added on certain sites.
 * Fixing segfault when web-inspector is closed
 * Fixing several memory leaks.
 * Fixing a use-after-free and a too small malloc
 * Don't unescape uri in the statusbar if it contains \n, \r or \f
 * Fixing issues with onscreen keyboards
 * Fixing issues with autocompletion


Changes since 2012-02-01 

New features:
  * New settings 'close-tab-focus-policy', 'new-tab-position-policy''
  * New setting 'ssl-use-system-ca-file' for libsoup > 2.38
  * New commands 'hints_background' 'hints_images_background', 'hints_url_background', opens hints in a background tab
  * New setting 'update-search-delay'
  * New commands 'back_open', 'backopen_url'
  * Adding case sensitive search
  * Tilde expansion for downloads and pathcompletion
  * New command 'dump'
  * New commands 'toggle_local_setting', 'local_set'
  * New settings 'font-hidden-statusbar'
  * New setting 'hint-auto-follow'
  * New command 'normal_mode'
  * New setting 'use-ntlm'
  * New dwb-uri 'dwb:searchengines'
  * New command 'cancel_download'
  * New command 'clear_tab'
  * New settings 'download-gradient-start', 'download-gradient-end'
  * New setting 'auto-insert-mode'
  * New command 'reload_userscripts'
  * New commands 'tab_move', 'tab_move_left', 'tab_move_right'
  * Adding 'DWB_MIME_TYPE' environment variable for external download scripts
  * New setting 'tabbar-visible'
  * Adding to small status bars if the statusbar is hidden that show links under
    pointer and insert mode state 

Changed: 
  * All dwb://<name> uris are replaced with dwb:<name>
  * Additional dbus backend for single instance handling
  * Possibility to send commands in userscripts to stdout, must be sent to
    $DWB_FIFO instead

Removed: 
  * Removed settings 'mail-handler', 'ftp-handler'

Bug fixes: 
  * Fixing broken downloads with libwebkit > 1.8.0
  * Fixing a possible segfault in the proxy setting
  * Fixing several memory leaks
  * Fixing not saved mimetypes for downloads opened with external applications
  * Fixing segfault when startpage is not set
  * Fixing several typos 
  * Several minor bug fixes



scripts/lib/dwb.js
(function() {
    var _modules = {};
    var _requires = {};
    var _callbacks = [];
    var _initialized = false;
    var _applyRequired = function(names, callback) 
    {
        if (names === null) 
            callback.call(this, _modules);
        else 
        {
            var modules = [];
            var name, detail;
            for (var j=0, m=names.length; j<m; j++) 
            {
                name = names[j];
                if (/^\w*!/.test(name)) 
                {
                    detail = name.split("!");
                    name = detail[0];
                    if (!_modules[name]) 
                        include(detail.slice(1).join("!"));
                }
                modules.push(_modules[name]);
            } 
            if (callback)
                callback.apply(this, modules);
        }
    };

    var _privProps = [];
    var _getPrivateIdx = function(object, key, identifier) 
    {
        var p;
        for (var i=0, l=_privProps.length; i<l; ++i) 
        {
            p = _privProps[i];
            if (p.object == object && p.key == key && p.identifier === identifier)
                return i;
        }
        return -1;
    };
    var _contexts;

    Object.defineProperties(this, { 
            "provide" : 
            { 
                value : function(name, module) 
                {
                    if (_modules[name]) 
                    {
                        io.debug({ 
                                offset : 1, arguments : arguments,
                                error : new Error("provide: Module " + 
                                                  name + " is already defined!")
                        });
                    }
                    else 
                        _modules[name] = module;
                }
            },
            "replace" : 
            {
                value : function(name, module) 
                {
                    if (! module && _modules[name] ) 
                    {
                        for (var key in _modules[name]) 
                            delete _modules[name][key];

                        delete _modules[name];
                    }
                    else 
                        _modules[name] = module;
                }
            },
            "require" : 
            {
                value : function(names, callback) 
                {
                    if (names !== null && ! (names instanceof Array)) 
                    {
                        io.debug({ 
                                error : new Error("require : invalid argument (" + 
                                                    JSON.stringify(names) + ")"), 
                                offset : 1, 
                                arguments : arguments 
                        });
                        return; 
                    }

                    if (!_initialized) 
                        _callbacks.push({callback : callback, names : names});
                    else 
                        _applyRequired(names, callback);
                }
            },
            // Called after all scripts have been loaded and executed
            // Immediately deleted from the global object, so it is not callable
            // from other scripts
            "_initAfter" : 
            { 
                value : function() 
                {
                    _initialized = true;
                    for (var i=0, l=_callbacks.length; i<l; i++) 
                        _applyRequired(_callbacks[i].names, _callbacks[i].callback);

                },
                configurable : true
            },
            "_initBefore" : 
            {
                value : function(contexts) 
                {
                    _contexts = contexts;
                    Object.freeze(this);
                },
                configurable : true
            }
    });
    Object.defineProperties(GObject.prototype, {
            "setPrivate" : 
            { 
                value : function(key, value, identifier) 
                {
                    if (!(identifier instanceof Object) && !(identifier instanceof Function)) 
                        throw new Error("[setPrivate] identifier is not an Object or Function");

                    var i = _getPrivateIdx(this, key, identifier);
                    if (i === -1) 
                    {
                        if (value !== undefined && value !== null)
                            _privProps.push({ 
                                    object : this, 
                                    key : key, 
                                    identifier : identifier, 
                                    value : value 
                            });
                    }
                    else if (value !== null) 
                    {
                        _privProps[i].value = value;
                    }
                    else 
                    {
                        _privProps.splice(i);
                    }
                }
            },
            "getPrivate" : 
            { 
                value : function(key, identifier) 
                {
                    var i = _getPrivateIdx(this, key, identifier);
                    if (i !== -1) 
                        return _privProps[i].value;
                    return undefined;
                }
            },
            "notify" : 
            { 
                value : function(name, callback, after) 
                { 
                    return this.connect("notify::" + util.uncamelize(name), callback, after || false);
                }
            },
            "connectBlocked" : 
            { 
                value : function(name, callback, after) 
                { 
                    var sig = this.connect(name, (function() { 
                        this.blockSignal(sig);
                        var result = callback.apply(callback, arguments);
                        this.unblockSignal(sig);
                        return result;
                    }).bind(this));
                    return sig;
                }
            },
            "notifyBlocked" : 
            {
                value : function(name, callback, after) 
                {
                    return this.connectBlocked("notify::" + util.uncamelize(name), callback, after || false);
                }
            }
    });
    Object.defineProperties(Deferred.prototype, {
            "done" : {
                value : function(method) {
                    return this.then(method);
                }
            },
            "fail" : {
                value : function(method) {
                    return this.then(null, method);
                }
            },
            "always" : {
                value : function(method)
                {
                    return this.then(method, method);
                }
            }
    });
})();
Object.preventExtensions(this);
